packages/plugs/lib/settings_page.ts

```
@@ -31,3 +35,24 @@ export async function readSettings<T extends object>(settings: T): Promise<T> {
     throw e;
   }
 }
+
+/**
+ * Convenience function to write a specific set of settings from the `SETTINGS` page.
+ * If the SETTiNGS page doesn't exist it will create it.
+ * @param settings 
+ */
+export async function writeSettings<T extends object>(settings: T) {
+  let readSettings = {};
+  try {
+    readSettings = (await readYamlPage(SETTINGS_PAGE, ["yaml"])) || {};
+  } catch (e: any) {
+    console.log("Couldn't read settings, generating a new settings page");
+    flashNotification("Creating a new SETTINGS page...", "info");
```

I wouldn't put UI stuff in something generally usable (from both client and server). It will break when using this API in a server context (there's no implementation of `flashNotification` there). Since you are not using `await` here it will silently fail (I think), but still.

---

> there's no implementation of flashNotification there.

I still think it is important to let the user know that something has gone wrong. In this one is about creating a new settings page so user is not surprised and could be removed (although I prefer to let them know that there is a new page in their space). But a couple of lines later I use the same mechanism for error/success. if a setting fails to update I'd like the user to know.

is there a way to tell if this is client/server code? and once we know we are in the server, how do we tell the user about anything happening? (other than throwing errors)

---

There’s no mechanism right now to know if you’re being run on the client or server (although you can mark a function with `env: server` in the yaml file to make sure that it does only run in one env). What I usually do is throw an exception, these exceptions transfer across client-server boundaries. So in an API like settings_page, I’d simply throw an exception when things go wrong.

Then on the client side I can catch this exception, even if it occurs on the server when I invoke it remotely via `invokeFunction(“server”, …)` I can put a try-catch around it on the client side, and use a `flashNotification` in the catch-clause to show the error.